jax_omeroutils/xml_editor.py

```diff
@@ -0,0 +1,234 @@
+    max_ann = 0
+    for ann in newome.structured_annotations:
+        clean_id = int(ann.id.split(":")[-1])
+        if clean_id > max_ann:
+            max_ann = clean_id
+    ann_count = max_ann + 1
+    for line in md:
+        filename = line['filename']
+        ann_dict = {i: line[i] for i in columns}
+        ann_dict.pop('filename')
+        ann_dict = {k: v for k, v in ann_dict.items() if isinstance(v, str)}
+        for ann in ome.structured_annotations:
+            if isinstance(ann, CommentAnnotation):
+                src_file = ann.value.split(sep)[-1]
+                if src_file == filename:
+                    img_id = ann.namespace
+                    for i in newome.images:
+                        if i.id == img_id:
+                            mmap = []
+                            for _key, _value in ann_dict.items():
+                                if _value:
+                                    mmap.append(M(k=_key, value=str(_value)))
+                                else:
+                                    mmap.append(M(k=_key, value=''))
+                            ann, annref = create_kv_and_ref(
+                                id=f"Annotation:{ann_count}",
+                                namespace=CURRENT_MD_NS,
+                                value=Map(ms=mmap))
+                            ann_count += 1
+                            newome.structured_annotations.append(ann)
+                            i.annotation_ref.append(annref)
+    return newome
+
+
+def add_annotations_plates(ome, imp_json):
+    newome = copy.deepcopy(ome)
+    columns = list(imp_json['user_supplied_md']['file_metadata'][0].keys())
+    if 'screen' in columns:
+        columns.remove('screen')
+    md = imp_json['user_supplied_md']['file_metadata']
+    max_ann = 0
+    for ann in newome.structured_annotations:
+        clean_id = int(ann.id.split(":")[-1])
+        if clean_id > max_ann:
+            max_ann = clean_id
+    ann_count = max_ann + 1
+    for line in md:
+        filename = line['filename']
+        ann_dict = {i: line[i] for i in columns}
+        ann_dict.pop('filename')
+        ann_dict = {k: v for k, v in ann_dict.items() if isinstance(v, str)}
+        for ann in ome.structured_annotations:
+            if isinstance(ann, CommentAnnotation):
+                src_file = ann.value.split(sep)[-1]
+                if src_file == filename:
+                    pl_id = ann.namespace
+                    for pl in newome.plates:
+                        if pl.id == pl_id:
+                            mmap = []
+                            for _key, _value in ann_dict.items():
+                                if _value:
+                                    mmap.append(M(k=_key, value=str(_value)))
+                                else:
+                                    mmap.append(M(k=_key, value=''))
+                            ann, annref = create_kv_and_ref(
```

Comment:
Same thing here: ann variable from the loop is overwritten, which makes it a bit hard to parse

Reply:
these loops are horrendous regardless, but agreed